In [2]:
farm = {
    '이수민': {
        '개운죽': 4.0,
        '금전수': 5.0,
        '떡갈잎고무나무': 3.0,
        '몬스테라': 1.5,
        '무늬접란': 1.5,
    },
    '백승진': {
        '개운죽': 5.0,
        '금전수': 4.0,
        '떡갈잎고무나무': 3.5,
        '몬스테라': 1.5,
        '무늬접란': 1.0 ,
        
    },
    
    '장지선': {
        '개운죽': 5.0,
        '금전수': 4.5,
        '떡갈잎고무나무': 2.5,
        '몬스테라': 1.0,
        '무늬접란': 1.0,
        
    },
    '조현빈': {
        '개운죽': 3.5,
        '금전수': 5.0,
        '떡갈잎고무나무': 3.0,
        '몬스테라': 1.5,
        '무늬접란': 1.0,
        
    },
    '홍성목': {
        '개운죽': 4.0,
        '금전수': 4.5,
        '떡갈잎고무나무': 3.0,
        '몬스테라': 1.0,
        '무늬접란': 2.5,
        
    },
    '김지현': {
        '개운죽': 4.0,
        '금전수': 4.0,
        '떡갈잎고무나무': 3.0,
        '몬스테라': 1.5,
        '무늬접란': 2.0,
        
    },
    '이태희': {
        '개운죽': 4.0,
        '금전수': 4.0,
        '떡갈잎고무나무': 2.5,
        '몬스테라': 1.0,
        '무늬접란': 2.0,
        
    },
    '박현정': {
        '개운죽': 4.0,
        '금전수': 0,
        '떡갈잎고무나무': 0,
        '몬스테라': 0,
        '무늬접란': 0, # 1. 아직 키우지 않는다 = 평가되지 않음?!! 2. 선호하지 않는다
    },
}

# 개운죽을 좋아하는 박현정
# 다른사람들도 개운죽을 좋아하면서, 금전수, 떡갈잎고무나무를 좋아함 (선호도: 금전수 > 떡갈잎고무나무)

In [3]:
import pandas as pd
import numpy as np
from collections import defaultdict
from scipy.stats.stats import pearsonr

In [4]:
# 유사도 테이블 생성
n = len(farm)
zeros = np.zeros((n, n))
sim = pd.DataFrame(zeros, index=list(farm.keys()), columns=list(farm.keys()))
sim

이수민  백승진  장지선  조현빈  홍성목  김지현  이태희  박현정
이수민  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0
백승진  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0
장지선  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0
조현빈  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0
홍성목  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0
김지현  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0
이태희  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0
박현정  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0

In [5]:
farm = pd.DataFrame(farm).T
farm

개운죽  금전수  떡갈잎고무나무  몬스테라  무늬접란
이수민  4.0  5.0      3.0   1.5   1.5
백승진  5.0  4.0      3.5   1.5   1.0
장지선  5.0  4.5      2.5   1.0   1.0
조현빈  3.5  5.0      3.0   1.5   1.0
홍성목  4.0  4.5      3.0   1.0   2.5
김지현  4.0  4.0      3.0   1.5   2.0
이태희  4.0  4.0      2.5   1.0   2.0
박현정  4.0  0.0      0.0   0.0   0.0

In [6]:
# 유사도 계산
for name1 in sim.index:
    for name2 in sim.index:
        sim[name1].loc[name2] = round(pearsonr(list(farm.loc[name1]), list(farm.loc[name2]))[0], 4) # scipy 라이브러리 사용
sim

이수민     백승진     장지선     조현빈     홍성목     김지현     이태희     박현정
이수민  1.0000  0.8850  0.9438  0.9857  0.9181  0.9604  0.9331  0.3627
백승진  0.8850  1.0000  0.9553  0.8499  0.8076  0.9375  0.8764  0.6594
장지선  0.9438  0.9553  1.0000  0.8899  0.8932  0.9741  0.9583  0.6504
조현빈  0.9857  0.8499  0.8899  1.0000  0.8533  0.9087  0.8603  0.2439
홍성목  0.9181  0.8076  0.8932  0.8533  1.0000  0.9608  0.9802  0.4082
김지현  0.9604  0.9375  0.9741  0.9087  0.9608  1.0000  0.9838  0.5393
이태희  0.9331  0.8764  0.9583  0.8603  0.9802  0.9838  1.0000  0.5574
박현정  0.3627  0.6594  0.6504  0.2439  0.4082  0.5393  0.5574  1.0000

In [7]:
sorted(list(dict(sim['이수민']).items()), key = lambda x:x[1], reverse=True)

[('이수민', 1.0),
 ('조현빈', 0.9857),
 ('김지현', 0.9604),
 ('장지선', 0.9438),
 ('이태희', 0.9331),
 ('홍성목', 0.9181),
 ('백승진', 0.885),
 ('박현정', 0.3627)]

In [8]:
# 새로운 데이터 추가
def update_table(update_name, new_data, new_person = False):
    if new_person:
        sim[update_name] = 1
        sim.loc[update_name] = 1
    
    farm.loc[update_name] = new_data

    for name in sim.index:
        sim[update_name].loc[name] = round(pearsonr(list(farm.loc[update_name]), list(farm.loc[name]))[0], 4)
        sim[name].loc[update_name] = round(pearsonr(list(farm.loc[update_name]), list(farm.loc[name]))[0], 4)

new_data = {
        '개운죽': 5.0,
        '금전수': 5.0,
        '떡갈잎고무나무': 4.0,
        '몬스테라': 1.5,
        '무늬접란': 1.5,
    }
update_table('이승형', new_data, new_person = True)

C:\Users\SSAFY\AppData\Local\Temp\ipykernel_17204\3473708329.py:10: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  sim[update_name].loc[name] = round(pearsonr(list(farm.loc[update_name]), list(farm.loc[name]))[0], 4)


In [9]:
farm

개운죽  금전수  떡갈잎고무나무  몬스테라  무늬접란
이수민  4.0  5.0      3.0   1.5   1.5
백승진  5.0  4.0      3.5   1.5   1.0
장지선  5.0  4.5      2.5   1.0   1.0
조현빈  3.5  5.0      3.0   1.5   1.0
홍성목  4.0  4.5      3.0   1.0   2.5
김지현  4.0  4.0      3.0   1.5   2.0
이태희  4.0  4.0      2.5   1.0   2.0
박현정  4.0  0.0      0.0   0.0   0.0
이승형  5.0  5.0      4.0   1.5   1.5

In [10]:
sim

이수민     백승진     장지선     조현빈     홍성목     김지현     이태희     박현정     이승형
이수민  1.0000  0.8850  0.9438  0.9857  0.9181  0.9604  0.9331  0.3627  0.9559
백승진  0.8850  1.0000  0.9553  0.8499  0.8076  0.9375  0.8764  0.6594  0.9723
장지선  0.9438  0.9553  1.0000  0.8899  0.8932  0.9741  0.9583  0.6504  0.9572
조현빈  0.9857  0.8499  0.8899  1.0000  0.8533  0.9087  0.8603  0.2439  0.9312
홍성목  0.9181  0.8076  0.8932  0.8533  1.0000  0.9608  0.9802  0.4082  0.8966
김지현  0.9604  0.9375  0.9741  0.9087  0.9608  1.0000  0.9838  0.5393  0.9783
이태희  0.9331  0.8764  0.9583  0.8603  0.9802  0.9838  1.0000  0.5574  0.9254
박현정  0.3627  0.6594  0.6504  0.2439  0.4082  0.5393  0.5574  1.0000  0.5020
이승형  0.9559  0.9723  0.9572  0.9312  0.8966  0.9783  0.9254  0.5020  1.0000

In [11]:
new_data = {
        '개운죽': 5.0,
        '금전수': 5.0,
        '떡갈잎고무나무': 5.0,
        '몬스테라': 5.5,
        '무늬접란': 5.0,
    }
update_table('이수민', new_data, new_person = False)

In [12]:
farm

개운죽  금전수  떡갈잎고무나무  몬스테라  무늬접란
이수민  5.0  5.0      5.0   5.5   5.0
백승진  5.0  4.0      3.5   1.5   1.0
장지선  5.0  4.5      2.5   1.0   1.0
조현빈  3.5  5.0      3.0   1.5   1.0
홍성목  4.0  4.5      3.0   1.0   2.5
김지현  4.0  4.0      3.0   1.5   2.0
이태희  4.0  4.0      2.5   1.0   2.0
박현정  4.0  0.0      0.0   0.0   0.0
이승형  5.0  5.0      4.0   1.5   1.5

In [13]:
sim

이수민     백승진     장지선     조현빈     홍성목     김지현     이태희     박현정     이승형
이수민  1.0000 -0.4945 -0.5322 -0.4529 -0.8165 -0.6864 -0.7289 -0.2500 -0.5961
백승진 -0.4945  1.0000  0.9553  0.8499  0.8076  0.9375  0.8764  0.6594  0.9723
장지선 -0.5322  0.9553  1.0000  0.8899  0.8932  0.9741  0.9583  0.6504  0.9572
조현빈 -0.4529  0.8499  0.8899  1.0000  0.8533  0.9087  0.8603  0.2439  0.9312
홍성목 -0.8165  0.8076  0.8932  0.8533  1.0000  0.9608  0.9802  0.4082  0.8966
김지현 -0.6864  0.9375  0.9741  0.9087  0.9608  1.0000  0.9838  0.5393  0.9783
이태희 -0.7289  0.8764  0.9583  0.8603  0.9802  0.9838  1.0000  0.5574  0.9254
박현정 -0.2500  0.6594  0.6504  0.2439  0.4082  0.5393  0.5574  1.0000  0.5020
이승형 -0.5961  0.9723  0.9572  0.9312  0.8966  0.9783  0.9254  0.5020  1.0000

In [14]:
def get_recommendation(name):
    plants =list(farm.columns)
    score_dic=defaultdict(int) # 평점
    sim_dic=defaultdict(int) # 유사도

    for name2, sim_score in dict(sim[name]).items():
        if sim_score > 0 and name != name2: # 박현정과 유사도가 0 이상인 사람에 대해서
            for plant in plants: # 모든 식물을 조사해본다.
                score = 0
                if not farm.loc['박현정'][plant]: # 만약 박현정이 그 식물을 키우지 않는다면,
                    score += farm.loc[name2][plant] * sim_score # 평점 예측 
                    score_dic[plant] += score
                    sim_dic[plant] += sim_score
    
    print(score_dic)
    print(sim_dic)
    ret = []
    for key in score_dic.keys():
        score_dic[key] = score_dic[key]/sim_dic[key]
        ret.append([score_dic[key], key])
    
    return ret

get_recommendation('박현정')

defaultdict(<class 'int'>, {'금전수': 15.517599999999998, '떡갈잎고무나무': 10.909600000000001, '몬스테라': 4.5329, '무늬접란': 5.5206})
defaultdict(<class 'int'>, {'금전수': 3.5606, '떡갈잎고무나무': 3.5606, '몬스테라': 3.5606, '무늬접란': 3.5606})


[[4.358141886198955, '금전수'],
 [3.0639779812391175, '떡갈잎고무나무'],
 [1.2730719541650284, '몬스테라'],
 [1.550469022074931, '무늬접란']]

# score_dic: 금전수: 15.5, 떡갈잎고무나무: 10.9, 몬스테라: 4.5, 무늬접란: 5.5
# sim_dic: 모든 식물 모두 3.5로 동일

# 데이터 생성시에 의도한대로 금전수가 1등, 떡갈잎이 2등